migrations/v0.5.6/document-detach.go

```diff
@@ -0,0 +1,188 @@
+
+	cursor, err := collection.Find(ctx, filter)
+	if err != nil {
+		return nil, err
+	}
+
+	for cursor.Next(ctx) {
+		var info database.ClientInfo
+		if err := cursor.Decode(&info); err != nil {
+			return nil, fmt.Errorf("decode client info: %w", err)
+		}
+
+		// 01. ensure deactivated
+		if info.Status != database.ClientDeactivated {
+			continue
+		}
+
+		// 02. ensure detached
+		hasAttachedDocs := false
+		for _, clientDocInfo := range info.Documents {
+			if clientDocInfo.Status == database.DocumentAttached {
+				hasAttachedDocs = true
+				break
+			}
+		}
+		if hasAttachedDocs {
+			failedClients = append(failedClients, info.ID.String())
+		}
+	}
+
+	return failedClients, nil
+}
+
+// processMigrationBatch processes the migration batch.
+func processMigrationBatch(
+	ctx context.Context,
+	collection *mongo.Collection,
+	infos []*database.ClientInfo,
+) {
+	for _, info := range infos {
+		result := collection.FindOneAndUpdate(ctx, bson.M{
+			"project_id": info.ProjectID,
+			"_id":        info.ID,
+		}, bson.M{
+			"$set": bson.M{
+				StatusKey:    database.ClientActivated,
+				"updated_at": time.Now().AddDate(-1, 0, 0),
+			},
+		})
+		if result.Err() != nil {
+			if errors.Is(result.Err(), mongo.ErrNoDocuments) {
+				_ = fmt.Errorf("%s: %w", info.Key, database.ErrClientNotFound)
+			} else {
+				_ = fmt.Errorf("update client info: %w", result.Err())
+			}
+		}
+	}
+}
+
+// ReactivateClients migrates the client collection to activate the clients
+// that are in a deactivated but have attached documents.
+func ReactivateClients(
+	ctx context.Context,
+	db *mongo.Client,
+	databaseName string,
+	batchSize int,
+) error {
+	collection := db.Database(databaseName).Collection("clients")
+
+	filter := bson.M{
+		"status": "deactivated",
+		"documents": bson.M{
+			"$ne":     nil,
+			"$exists": true,
+		},
+		"$expr": bson.M{
+			"$gt": bson.A{
+				bson.M{
+					"$size": bson.M{
+						"$filter": bson.M{
+							"input": bson.M{
+								"$objectToArray": "$documents",
+							},
+							"as": "doc",
+							"cond": bson.M{
+								"$eq": bson.A{"$$doc.v.status", "attached"},
+							},
+						},
+					},
+				},
+				0,
+			},
+		},
+	}
+
+	totalCount, err := collection.CountDocuments(ctx, filter)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("total clients: %d\n", totalCount)
+
+	cursor, err := collection.Find(ctx, filter)
+	if err != nil {
+		return err
+	}
+
+	var infos []*database.ClientInfo
+
+	batchCount := 1
+	done := 0
+	for cursor.Next(ctx) {
+		var clientInfo database.ClientInfo
+		if err := cursor.Decode(&clientInfo); err != nil {
+			return fmt.Errorf("decode client info: %w", err)
+		}
+
+		infos = append(infos, &clientInfo)
+
+		if len(infos) >= batchSize {
+			processMigrationBatch(ctx, collection, infos)
+			done += len(infos)
+
+			percentage := int(float64(batchSize*batchCount) / float64(totalCount) * 100)
+			fmt.Printf("%s.clients migration progress: %d%%(%d/%d)\n", databaseName, percentage, done, totalCount)
```

Comment:
> _⚠️ Potential issue_
> 
> **Prevent division by zero when calculating progress percentage**
> 
> When `totalCount` is zero, the calculation of `percentage` results in a division by zero error.
> 
> Add a check to prevent division by zero and correct the calculation:
> 
> ```diff
> - percentage := int(float64(batchSize*batchCount) / float64(totalCount) * 100)
> + var percentage int
> + if totalCount > 0 {
> +     percentage = int(float64(done) / float64(totalCount) * 100)
> + } else {
> +     percentage = 100
> + }
> ```
> 
> 📝 Committable suggestion

If the result of the previous countDocument query is not 0, shouldn't cursor.Next(ctx) be nil? In other words, wouldn't the fact that cursor.Next(ctx) is not nil guarantee that totalCount is not 0?